<div style="text-align: right">
    <i>
        LIN 537: Computational Lingusitics 1 <br>
        Fall 2019 <br>
        Alëna Aksënova
    </i>
</div>

# Notebook 9: classes and inheritance

This notebook covers the notion of a `class` and introduces the notion of **object-oriented programming** (OOP) in comparison to the **functional** programming. As of the Python-specific knowledge, this notebook exemplifies how to define a new class, its attributes and its methods. It also shows how to inherit attributes and methods from parent classes.

We will focus on two **programming paradigms:** _functional_ programming and _OOP._
Roughly speaking, **functional** programming implements the _actions,_ or functions, that map some input to some output. **Object-oriented** programming instead defines objects, and then describes how these objects can be modified. Intuitively: focus on the functions/procedures vs. focus on the properties of the objects.

Before we learned how to define custom functions using `def` or `lambda`, and it is possible to define custom classes as well.

## Class definition

A **class** is a representation of an object. For example, this object can have the following properties:

  * name: Car
  * attributes: Car.make (i.e. Jeep), Car.color (i.e. black), Car.year (i.e. 2006)
  * methods: Car.get_fuel(), Car.drive(speed), Car.lock()
  
In Python, these properties (attributes and methods) need to be listed in the _class definition._ In order to define a new class, we use the `class` operator followed by a name of the class.

    class NewObject(object):
        # code
        
In the parenthesis after the name of the class we list classes from which some functionality will be _inherited,_ but if none of such classes exist, the conventional way is to write `object` there. It simply means that we are instantiating an object.

In [0]:
class Rectangle(object):
    
    # operator that means "do nothing"
    pass

## Attributes definition

The next step will be to list properties, or **attributes,** of that object, i.e. color, size, make, etc. Attributes do not refer to any actions, they simply describe the features of that object.

    class NewObject(object):
    
        def __init__(self, make, year):
            self.make = make
            self.year = year
            
The function `__init__` _always_ must be present in the class definition: it will instantiate properties of that class. Notice the `self` argument of this function: it basically means that this function, or _method,_ is operating on the class itself. `__init__` initializes the attributes of the class `NewObject`.

In [0]:
class Rectangle(object):
    
    def __init__(self, side_1, side_2, color):
        self.side_1 = side_1
        self.side_2 = side_2
        self.color = color

Here, we are initializing a class `Rectangle`, and it will have 3 attributes: `side_1`, `side_2` and `color`. Whatever arguments `__init__` takes, except `self`, must be provided as arguments of the class itself upon initialization.

In [0]:
x = Rectangle(5, 3, "red")

Now when the class is initialized, we can access its attributes directly. **Dot operator** is used to access an attribute or a method of a particular class:
    
    class_name.attribute_name
    class_name.method_name(arg1, arg2)

In [6]:
print(x.color)

red


Notice that the names of the arguments of `__init__` and the names of the attributes don't need to match, but it is a convention to have them matching.

In [7]:
class A(object):
    def __init__(self, some_color, some_year):
        self.color = some_color
        self.year = some_year
        
obj = A("red", 2004)
print("Color:", obj.color)
print("Year:", obj.year)

Color: red
Year: 2004


The value of the attributes can be changed directly. (The situation will become more complicated if we are working with _protected_ values.)

In [8]:
print("Old side:", x.side_2)
x.side_2 = 10
print("New side:", x.side_2)

Old side: 3
New side: 10


If we have more that one object defined, dot operator helps us not to get confused in attributes of different objects:

In [9]:
a = Rectangle(5, 3, "red")
b = Rectangle(1, 10, "blue")

print("Color or A:", a.color)
print("Color or B:", b.color)

Color or A: red
Color or B: blue


**Practice:** set up the classes for the following objects: `Book`, `Door`, `Cat`.

## Methods definition

Now let us define some methods. They look like functions defined inside classes. The crucial difference is that methods must have `self` as the first argument: it simply means that that function/method is applied to the object itself.

    class NewObject(object):
    
        def __init__(self, make, year):
            self.make = make
            self.year = year
            
        def method_name(self, arg1, arg2):
            # code

In [0]:
class Rectangle(object):
    
    def __init__(self, side_1, side_2, color):
        self.side_1 = side_1
        self.side_2 = side_2
        self.color = color
        
    def calculate_area(self):
        return self.side_1 * self.side_2

The method `calculate_area` in the code above requires no arguments apart from `self` because the information about the sides does not need to be provided: _it is already available!_ Calling `self.side_1` and `self.side_2` allows us to get that information.

In [11]:
a = Rectangle(5, 3, "red")
a.calculate_area()

15

The method `calculate_area` calculates the area of the rectangle and returns the value. However, we can instead go ahead and save the value into a special attribute `area`. Notice that we do not expect `area` to be provided beforehand. **Every attribute, even if its value is not known yet, needs to be initialized inside the `__init__`.**

In [0]:
class Rectangle(object):
    
    def __init__(self, side_1, side_2, color):
        self.side_1 = side_1
        self.side_2 = side_2
        self.color = color
        self.area = None
        
    def calculate_area(self):
        self.area = self.side_1 * self.side_2

In [14]:
a = Rectangle(5, 3, "red")
a.calculate_area()
print(a.area)

15


In [15]:
a.area

15

**Question:** is it possible for the `area` attribute to contain a wrong value?

**Practice:** add a method `is_square` to the class `Rectangle` that will return `True` if the rectangle is a square, and `False` otherwise.

In [5]:
class Rectangle(object):
    
    def __init__(self, side_1, side_2, color):
        self.side_1 = side_1
        self.side_2 = side_2
        self.color = color
        self.area = None
        
    def calculate_area(self):
        self.area = self.side_1 * self.side_2
        
    # add the code for the method "is_square" here
    def is_square(self):
      if self.side_1==self.side_2:
        return True
      else:
        return False
a=Rectangle(5,5,"red")
a.is_square()

True

In [0]:
'''def is_square(self):
  if self.side_1==self.side_2:
    return True  (same as the above)'''

def is_square(self):
  return self.side_1==self.side_2


## Inheritance

Remember that every class upon definition inherits from the `object`:

    class Rectangle(object):
        # code
        
In a similar way, classes can _inherit_ from other classes as well:

    class SomeClass(ItsParentClass):
        # code
        
Let's say we want to define a class `Vehicle` implementing generic methods and attrbutes for vehicles, and then we want to introduce classes `Boat` and `Car` that inherit from `Vehicle` and add additional car-specific or boat-specific functionality.

Let us first implement the class `Vehicle`.

In [0]:
class Vehicle(object):
    """ Class implementing a Vehicle object. """
    
    def __init__(self, year, max_speed):
        """ Initializes the attributes year and max_speed. """
        self.year = year
        self.max_speed = max_speed
        
    def drive(self, speed):
        """ Safely drives the vehicle. """
        if speed <= self.max_speed:
            print("Driving! The speed is", speed, "mph.")
        else:
            print("Too fast! I'll be driving", self.max_speed, "mph.")

Test how the method `drive` behaves with different speeds.

In [18]:
jeep = Vehicle(2006, 70)

jeep.drive(99)

Too fast! I'll be driving 70 mph.


Now, let us initialize classes `Boat` and `Car` that will **inherit** from `Vehicle`.

In [0]:
class Boat(Vehicle):
    pass

class Car(Vehicle):
    pass

**Inheriting** from `Vehicle` means that all methods and attributes of the parental class are now available in the children classes as well.

In [20]:
jeep = Car(2006, 50)
print(jeep.year)

2006


In [21]:
titanic = Boat(1909, 0)
titanic.drive(30)

Too fast! I'll be driving 0 mph.


If you want to add some `Car`-specific methods, you can simply do so by simply defining new methods for that class. Not that **methods defined in the children classes are not available in the parent classes.**

In [22]:
class Car(Vehicle):
    """ A class defined for cars. """
    
    def pass_a_car(self, my_speed, their_speed):
        if my_speed > their_speed:
            print("I am passing that car.")
        else:
            print("I should not pass that car.")
            
jeep = Car(2006, 70)
jeep.pass_a_car(60, 50)
jeep.drive(80)

I am passing that car.
Too fast! I'll be driving 70 mph.


The function `pass_a_car` is indeed now available for the class `Car`, however, it is not defined for the sibling or parent classes.

In [24]:
lada = Vehicle(1985, 38)
lada.drive(39)

# The code below will cause an error!
# lada.pass_a_car(37, 60)

Too fast! I'll be driving 38 mph.


The method `pass_a_car` takes two arguments: `my_speed` and `their_speed`, where `my_speed` is the speed of the car itself. _Itself_ is a good indicator that we are dealing with a good candidate to become an attribute of the class `Car`. Let us add a new attribute to the class `Car`!

In [0]:
class Car(Vehicle):
    """ A class defined for cars. """
    
    def __init__(self, my_speed):
        self.my_speed = my_speed
    
    def pass_a_car(self, their_speed):
        if self.my_speed > their_speed:
            print("I am passing that car.")
        else:
            print("I should not pass that car.")

In fact, as the code below shows, `my_speed` is now an attribute.

In [26]:
car = Car(80)
car.pass_a_car(50)


I am passing that car.


However, now the new function `__init__` of the `Car` class rewrote the original `__init__` of the `Vehicle` completely: attributes `year` and `max_speed` are not implemented anymore.

In [0]:
# The code below will cause an error!
# car.year 

It we want to **add** add new atributes to the children class, we need to re-implement parent's attributes as well.

In [0]:
class Car(Vehicle):
    """ A class defined for cars. """
    
    def __init__(self, year, max_speed, my_speed):
        self.year = year
        self.max_speed = max_speed
        self.my_speed = my_speed
    
    def pass_a_car(self, their_speed):
        if self.my_speed > their_speed:
            print("I am passing that car.")
        else:
            print("I should not pass that car.")

Now, for the class `Car`, every attribute and method available in `Vehicle` is available as well, and also the newly defined attribute `my_speed` and the new method `pass_a_car`.

In [29]:
a = Car(2004, 80, 45)
print(a.year)
a.drive(45)
a.pass_a_car(70)

2004
Driving! The speed is 45 mph.
I should not pass that car.


If the `__init__` method of the class needs to be modified by adding attributes that are not implemented for the parent class, we can access and run `__init__` of the parent class by calling `super().__init__`.

    class Child(Parent):
    
        def __init__(self, parent_att1, parent_att2, new_att1):
            super().__init__(parent_att1, parent_att2)
            self.new_att1 = new_att1
            
The code below is the modified version of the previous definition of the class `Car`.

In [30]:
class Car(Vehicle):
    """ A class defined for cars. """
    
    def __init__(self, year, max_speed, my_speed):
        super().__init__(year, max_speed)
        self.my_speed = my_speed
    
    def pass_a_car(self, their_speed):
        if self.my_speed > their_speed:
            print("I am passing that car.")
        else:
            print("I should not pass that car.")
            
a = Car(2004, 80, 45)
print(a.year)
a.drive(45)
a.pass_a_car(70)

2004
Driving! The speed is 45 mph.
I should not pass that car.


If there are several classes and some of them inherit from others, the dependencies can quickly become complicated and hard to remember. To keep track of classes and the dependencies, people usually draw **class diagrams**. A simple class diagram representing the structure of the class `Vehicle` and its children classes is given below.

<img src="images/9_1.png" width="520">

**Practice:** implement the classes described in the class diagram below.

<img src="images/9_2.png" width="450">

In [0]:
class dog(Pet):
  def __init__(self,name,age,colors):
    self.name=name
    self.age=age
    self.colors= colors
  def get_older(self,years):
    self.age+= years
  def pet(self):
    print(self.name,"said \"I am happy now!\"")
  
  guido=Pet("Guido",)
  

In [0]:
class Dog(Pet):
  def init__(self,name,age,colors,tail_speed):
    super().__init__(name,age,colors)
    self.tail_speed=tail_speed
  def bark(self):
    print("Bow!")
guido=Dog("Guido",4,"blue"35)
guido.get_older(5)

# Homework 9

**Due on Sunday, November 3rd, 11.59pm**

Send your notebook (don't forget to save your solutions!) to <alena.aksenova@stonybrook.edu> with the subject **\[CompLing1\] Homework 9**.

### Problem: implementing classes for `Food` and `Liquid`.

By the end of the homework, you should implement two classes (`Food` and `Liquid`) described in the following class diagram.

<img src="images/9_3.png" width="450">

**Part 1: implementing the class `Food`.** 

It should have the following attributes:
  * `food_type`: "banana", "water", "zucchini", ...
  * `flavor`: "bitter", "sweet", ...
  * `quantity`: 0, 150, 5239, ...
  * `temperature`: 420, 375, 710, ...
  
It should also include the following methods:
  * `change_temperature(by_degrees)`: it changes the temperature attribute by the given number of the degrees;
  * `consume()`: it complains if the quantity attribute of the food is not positive, but otherwise it changes the quantity of the food to $0$ (like this we can show that the food is eaten).

In [0]:
class Food(object):
  def __init__(self,food_type,flavor,quantity,temperature):
    self.food_type=food_type
    self.flavor=flavor
    self.quantity=quantity
    self.temperature=temperature
  def change_temperature(self,degrees):
    if degrees>0:
      print("I heated up the food")
      print("The new temperature is",self.temperature+degrees,"degrees")
    elif degrees<0:
      print("I cooled down food")
      print("The new temperature is",self.temperature+degrees,"degrees")
  def consume(self):
    if len(self.food_type)>1:
      self.food_type=""
      print("I consumed all the food")
    else:
      print("There is no food")
  

In [65]:
borscht=Food("soup","salty",250,80)
print(borscht.flavor)

salty


In [66]:
borscht.change_temperature(5)

I heated up the food
The new temperature is 85 degrees


In [67]:
borscht.change_temperature(-15)

I cooled down food
The new temperature is 65 degrees


In [68]:
borscht.consume()

I consumed all the food


In [69]:
borscht.consume()

There is no food


Test the class `Food`.

**Example of the instructions and the expected behavior**

    borscht = Food("soup", "salty", 250, 80)
    print(borscht.flavor)
    >> salty

    borscht.change_temperature(5)
    >> I heated up the food.
    >> The new temperature is 85 degrees.
    
    borscht.change_temperature(-15)
    >> I cooled down the food.
    >> The new temperature is 70 degrees.

    borscht.consume()
    >> I consumed all the food.
    
    borscht.consume()
    >> There is no food!

In [0]:
# borscht = Food("soup", "salty", 250, 80)
# print(borscht.flavor)

# borscht.change_temperature(5)
# borscht.change_temperature(-15)

# borscht.consume()
# borscht.consume()

**Part 2: implementing the class `Liquid`.** 

Now, implement the class `Liquid` that will have also an attribute `boiling_temperature`, and the method `boil` that changes the `temperature` attribute to the boiling temperature. The class `Liquid` must inherit all other methods and attributes from the class `Food` as well.

In [0]:
class Liquid(Food):
  def __init__(self,food_type,flavor,quantity,temperature,boiling_temperature):
     super().__init__(food_type,flavor,quantity,temperature)
     self.boiling_temperature=boiling_temperature
  def boil(self):
    self.temperature=self.boiling_temperature
    print("Boiled the liquid! Its temperature now is ",self.temperature," degrees")
  

In [82]:
milk=Liquid("milk","creamy",16,180,212.3)
print(milk.boiling_temperature)

212.3


In [83]:
milk.change_temperature(-40)

I cooled down food
The new temperature is 140 degrees


In [84]:
milk.boil()

Boiled the liquid! Its temperature now is  212.3  degrees


In [85]:
milk.consume()

I consumed all the food


Test the class `Liquid`.

**Example of the instructions and the expected behavior**

    milk = Liquid("milk", "creamy", 16, 180, 212.3)
    print(milk.boiling_temperature)
    >> 212.3
    
    milk.change_temperature(-40)
    >> I cooled down the food.
    >> The new temperature is 140 degrees.

    milk.boil()
    >> Boiled the liquid! Its temperature now is 212.3 degrees.

    milk.consume()
    >> I consumed all the food.

In [0]:
# milk = Liquid("milk", "creamy", 16, 180, 212.3)
# print(milk.boiling_temperature)

# milk.change_temperature(-40)
# milk.boil()
# milk.consume()